crates/pg_statement_splitter/src/parser.rs

```diff
@@ -33,164 +37,94 @@ pub struct Parse {
 }
 
 impl Parser {
-    pub fn new(tokens: Vec<Token>) -> Self {
+    pub fn new(sql: &str) -> Self {
+        // we dont care about whitespace tokens, except for double newlines
+        // to make everything simpler, we just filter them out
+        // the token holds the text range, so we dont need to worry about that
+        let tokens = lex(sql)
+            .iter()
+            .filter(|t| {
+                return !WHITESPACE_TOKENS.contains(&t.kind)
+                    || (t.kind == SyntaxKind::Newline && t.text.chars().count() > 1);
+            })
+            .rev()
+            .cloned()
+            .collect::<Vec<_>>();
+
         Self {
-            eof_token: Token::eof(usize::from(tokens.last().unwrap().span.end())),
             ranges: Vec::new(),
+            eof_token: Token::eof(usize::from(tokens.first().unwrap().span.end())),
             errors: Vec::new(),
             current_stmt_start: None,
             tokens,
-            pos: 0,
-            whitespace_token_buffer: None,
+            last_token_end: None,
         }
     }
 
     pub fn finish(self) -> Parse {
         Parse {
-            ranges: self
-                .ranges
-                .iter()
-                .map(|(start, end)| {
-                    let from = self.tokens.get(*start);
-                    let to = self.tokens.get(end - 1);
-                    // get text range from token range
-                    let text_start = from.unwrap().span.start();
-                    let text_end = to.unwrap().span.end();
-
-                    TextRange::new(
-                        TextSize::try_from(text_start).unwrap(),
-                        TextSize::try_from(text_end).unwrap(),
-                    )
-                })
-                .collect(),
+            ranges: self.ranges,
             errors: self.errors,
         }
     }
 
-    pub fn start_stmt(&mut self) {
+    /// Start statement
+    pub fn start_stmt(&mut self) -> Token {
```

Comment:
Should we return a `Token` here? It might be confusing when a dev works with this and assumes that the next `.peek()` yields a different token

Reply:
yes, I am not happy with the start / stop stmt methods yet. they are doing things that you would not assume they would do. right now `start_stmt` `peek`s and sets the peeked tokens start as the current start statement. that feels off for me. do you think that not returning the token would improve this? I was thinking to maybe just have the user pass in the token on which to start as an argument.

Reply:
I refactored this a bit. see below :)